modules/apps/object/object-service/src/main/java/com/liferay/object/internal/sort/ObjectEntryMTo1RelationshipSortDSLQueryVisitor.java

```diff
@@ -0,0 +1,146 @@
+import com.liferay.portal.kernel.util.ListUtil;
+import com.liferay.portal.kernel.util.StringUtil;
+
+import java.util.Stack;
+
+/**
+ * @author Sergio Jiménez del Coso
+ */
+public class ObjectEntryMTo1RelationshipSortDSLQueryVisitor
+	extends BaseSortDSLQueryVisitor {
+
+	public ObjectEntryMTo1RelationshipSortDSLQueryVisitor(
+		ObjectFieldLocalService objectFieldLocalService,
+		ObjectRelationshipLocalService objectRelationshipLocalService) {
+
+		super(objectFieldLocalService, objectRelationshipLocalService);
+	}
+
+	public DSLQuery visit(DSLQuery dslQuery, Sort sort) throws PortalException {
+		RelationshipSort relationshipSort = (RelationshipSort)sort;
+
+		ObjectRelationship objectRelationship =
+			relationshipSort.getObjectRelationship();
+
+		ObjectDefinition relatedObjectDefinition =
+			relationshipSort.getRelatedObjectDefinition();
+
+		String dbColumnName = StringBundler.concat(
+			"r_", objectRelationship.getName(), "_",
+			relatedObjectDefinition.getPKObjectFieldName());
+
+		String formatedRelationshipPathName = StringUtil.replace(
+			relationshipSort.getFieldPath(), CharPool.FORWARD_SLASH,
+			CharPool.UNDERLINE);
+
+		DynamicObjectDefinitionTable relatedDynamicObjectDefinitionTable =
+			(DynamicObjectDefinitionTable)getAliasedTable(
+				formatedRelationshipPathName,
+				_toDynamicObjectDefinitionTable(relatedObjectDefinition));
+
+		if (!contains(dslQuery, relatedDynamicObjectDefinitionTable)) {
+			DynamicObjectDefinitionTable dynamicObjectDefinitionTable =
+				_getDynamicObjectDefinitionTable(
+					relationshipSort.getObjectDefinition(), dbColumnName);
+
+			if (!contains(dslQuery, dynamicObjectDefinitionTable)) {
+				dynamicObjectDefinitionTable =
+					(DynamicObjectDefinitionTable)getAliasedTable(
+						StringUtil.removeLast(
+							formatedRelationshipPathName,
+							"_" + objectRelationship.getName()),
+						dynamicObjectDefinitionTable);
+
+				dslQuery = addLeftJoin(
+					dynamicObjectDefinitionTable.getPrimaryKeyColumn(),
+					dslQuery, null, dynamicObjectDefinitionTable);
+			}
+
+			dslQuery = addLeftJoin(
+				(Column<DynamicObjectDefinitionTable, Long>)
+					dynamicObjectDefinitionTable.getColumn(dbColumnName),
+				dslQuery, relatedDynamicObjectDefinitionTable,
+				relatedDynamicObjectDefinitionTable);
+		}
+
+		Stack<BaseASTNode> allBaseASTNodes = getAllBaseASTNodes(
+			GroupByStep.class, dslQuery);
+
+		if (ListUtil.exists(allBaseASTNodes, GroupBy.class::isInstance)) {
+			return dslQuery;
+		}
+
+		GroupByStep groupByStep = (GroupByStep)allBaseASTNodes.pop();
+
+		Stack<BaseASTNode> selectAllBaseASTNodes = getAllBaseASTNodes(
+			Select.class, dslQuery);
+
+		Select select = (Select)selectAllBaseASTNodes.pop();
+
+		BaseASTNode baseASTNode = (BaseASTNode)groupByStep.groupBy(
+			select.getExpressions(
+			).toArray(
+				new Expression[0]
+			));
+
+		return updateParents(baseASTNode, allBaseASTNodes);
+	}
+
+	private DynamicObjectDefinitionTable _getDynamicObjectDefinitionTable(
```

Comment:
You have created this private method but in the 1toM visitor we are getting the dynamic table from the dbColumnName with the following code:

```
objectFieldLocalService.getTable(
					relatedObjectDefinition.getObjectDefinitionId(),
					dbColumnName)
```

Is the method necessary? I am asking because I only see one invocation to this method, however I see an if clause inside, so it is not clear if both ways (inside the if and outside) are executed

Reply:
Yes, I've created this private method (at the moment) because this is useful is we have to use the normal table or the extension table. Remember that a Relationship Field can be in the normal or in the extension table depending when it is created. Maybe I can simplify this with a better approach and indicate which is the dbName (if it is the normal table or the extended table)

Reply:
As we discussed, I should use the service because it is implementing which type of table is is given the objectDefinition